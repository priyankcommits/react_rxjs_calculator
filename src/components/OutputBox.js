import React, { useEffect, useState } from 'react';
import { skip, delay } from 'rxjs/operators';

import mathService from '../services/mathService';
import { BoxStyled } from '../styles';
import { resultEvent$ } from '../events';

function OutputBox() {

  const [outputValue, setOutputValue] = useState('');
  const [previousAnswer, setPreviousAsnwer] = useState('');

  const executeMathFunction = ((data) => {
    const result = async () => {
      let response = await mathService(data, previousAnswer);
      if (response !== null) {
        setOutputValue(response);
        setPreviousAsnwer(response);
      }
    };
    result();
  });

  useEffect(() => {
    const resultEvent = resultEvent$.pipe(skip(1));
    resultEvent.pipe(delay(333)).subscribe(data => {
      executeMathFunction(data);
    });
  }, [previousAnswer]);

  return (
    <BoxStyled>
      {outputValue}
    </BoxStyled>
  )
}

export default OutputBox;
