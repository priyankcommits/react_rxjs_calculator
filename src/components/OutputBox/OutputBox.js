// Component to show output result and talks to math service
// Listens to InputBox operations and Publishes to History
import React, { useEffect, useState, useReducer } from 'react';
import { skip, delay } from 'rxjs/operators';

import { mathService } from '../../services/index';
import { BoxStyle, OverFlowProtect } from '../../shared/styles/main/index';
import { operationEvent$, resultEvent$ } from '../../shared/events/index';
import { initialOutputValue, outputReducer } from '../../reducers/index';

const OutputBox = () => {

  const [outputValue, setOutputValue] = useState('');
  const [output, setOutputReducer] = useReducer(outputReducer, initialOutputValue);

  useEffect(() => {
    const operationEvent = operationEvent$.pipe(skip(1));
    operationEvent.pipe(delay(333)).subscribe(data => {
      setOutputReducer(data);
    });
    return () => {
      operationEvent$.unsubscribe();
    }
  }, []);

  useEffect(() => {
    const waitResult = async () => {
      let response = await mathService(output, outputValue);
      if (!isNaN(response)) {
        setOutputValue(response);
        resultEvent$.next({input: output, value: response});
      }
    }
    waitResult();
  }, [output]);

  return (
    <BoxStyle>
      <OverFlowProtect>
        <h4>
          {outputValue}
        </h4>
      </OverFlowProtect>
    </BoxStyle>
  )
}

export { OutputBox };
