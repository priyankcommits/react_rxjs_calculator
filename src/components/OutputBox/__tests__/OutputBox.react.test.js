// Snapshot testing
import React from 'react';
import renderer from 'react-test-renderer';

import { OutputBox } from '../OutputBox';

test('Output Box Initial Rendering', () => {
  const component = renderer.create(
    <OutputBox />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});
