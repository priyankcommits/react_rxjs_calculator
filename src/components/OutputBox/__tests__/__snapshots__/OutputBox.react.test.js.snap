// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output Box Initial Rendering 1`] = `
<div
  className="sc-AxirZ bQmHBL"
>
  <div
    className="sc-AxiKw eGDepq"
  >
    <h4>
      
    </h4>
  </div>
</div>
`;
